libqrexec/exec.c

```diff
@@ -262,78 +252,205 @@ static int find_qrexec_service_file(
-        if (!realcmd) {
-            LOG(ERROR, "Bad command from dom0: no colon");
-            abort();
+        const char *colon = strchr(cmdline, ':');
+        if (!colon) {
+            LOG(ERROR, "Bad command from dom0 (%s): no colon", cmdline);
+            goto err;
         }
-        username = strndup(cmdline, (size_t)(realcmd - cmdline));
-        if (!username) {
+        cmd->username = strndup(cmdline, (size_t)(colon - cmdline));
+        if (!cmd->username) {
             PERROR("strndup");
-            abort();
+            goto err;
+        }
+        cmd->command = colon + 1;
+    } else
+        cmd->command = cmdline;
+
+    if (strncmp(cmd->command, NOGUI_CMD_PREFIX, NOGUI_CMD_PREFIX_LEN) == 0) {
+        cmd->nogui = true;
+        cmd->command += NOGUI_CMD_PREFIX_LEN;
+    } else
+        cmd->nogui = false;
+
+    /* If the command starts with "QUBESRPC ", parse service descriptor */
+    if (strncmp(cmd->command, RPC_REQUEST_COMMAND " ",
+                RPC_REQUEST_COMMAND_LEN + 1) == 0) {
+        const char *start, *end;
+
+        /* Parse service descriptor ("qubes.Service+arg") */
+
+        start = cmd->command + RPC_REQUEST_COMMAND_LEN + 1;
+        end = strchrnul(start, ' ');
+        if (!end) {
+            LOG(ERROR, "No space found after service descriptor");
+            goto err;
         }
-        realcmd++;
-    } else {
-        realcmd = cmdline;
-    }
 
-    // Get the part of the command line that will be executed.
-    const char *const start_cmdline = skip_nogui(realcmd);
-    if (strncmp(start_cmdline, RPC_REQUEST_COMMAND " ", RPC_REQUEST_COMMAND_LEN + 1) != 0) {
-        // Legacy qrexec behavior: spawn shell directly.
-        return do_fork_exec(username, realcmd, pid, stdin_fd, stdout_fd, stderr_fd);
-    } else {
-        // Proper Qubes RPC call
-        service_descriptor = start_cmdline + RPC_REQUEST_COMMAND_LEN + 1;
+        if (end - start > MAX_SERVICE_NAME_LEN) {
+            LOG(ERROR, "Command too long (length %zu)",
+                end - start);
+            goto err;
+        }
+
+        cmd->service_descriptor = strndup(start, end - start);
+        if (!cmd->service_descriptor) {
+            PERROR("strndup");
+            goto err;
+        }
+
+        /* Parse service name ("qubes.Service") */
+
+        const char *plus = memchr(start, '+', end - start);
+        if (plus) {
+            if (plus - start == 0) {
+                LOG(ERROR, "Service name empty");
+                goto err;
+            }
+            if (plus - start > NAME_MAX) {
+                LOG(ERROR, "Service name too long to execute (length %zu)",
+                    plus - start);
+                goto err;
+            }
+            cmd->service_name = strndup(start, plus - start);
+            if (!cmd->service_name) {
+                PERROR("strndup");
+                goto err;
+            }
+        } else {
+            cmd->service_name = strdup(cmd->service_descriptor);
```

Comment:
```suggestion
            cmd->service_name = strndup(cmd->service_descriptor, end - start);
```
?

Reply:
I'll do `strndup(start, end - start)`, same as for `cmd->service_descriptor`.